compiler/rustc_query_impl/src/plumbing.rs

```diff
@@ -91,19 +91,24 @@ impl QueryContext for QueryCtxt<'_> {
     fn start_query<R>(
         &self,
         token: QueryJobId,
+        depth_limit: bool,
         diagnostics: Option<&Lock<ThinVec<Diagnostic>>>,
         compute: impl FnOnce() -> R,
     ) -> R {
         // The `TyCtxt` stored in TLS has the same global interner lifetime
         // as `self`, so we use `with_related_context` to relate the 'tcx lifetimes
         // when accessing the `ImplicitCtxt`.
         tls::with_related_context(**self, move |current_icx| {
+            if depth_limit && !self.recursion_limit().value_within_limit(current_icx.query_depth) {
+                panic!("queries overflow the depth limit!");
```

Comment:
This used to be a fatal error message. Could you make it so?
Panics are considered ICEs, and ask for a bug report. This is not the case.

Reply:
Yea, I added `depth_limit_error` func in `QueryContext` so we can keep `#![deny(rustc::diagnostic_outside_of_impl)]` in `rustc_query_impl`